### ISSUE TITLE
Improved logic w/ input vs. output streams

### ISSUE THREAD

**Issue by jjlauer** [zeroturnaround/zt-exec#33]
For various reasons, I need to supply my own wrapped System.in inputstream.  The problem I'm running into is that the ExecuteStreamHandler makes a very important decision on what kind of InputStream pump it creates IF the inputstream is System.in.  In my case, the process runs and exits, but the InputStream is hanging forever on a blocking read() call.

I'm not sure this is the right logic for input streams vs. output.  In fact, if the process actually does exit, why aren't all the streams closed and their pumping threads interrupted?  Or if the output streams are closed, it seems like the inputstream w/o a doubt should be closed.

**Comment by jjlauer** [comment]
This is my workaround for now

```java
InputStream is = ....;

PumpStreamHandler streams = new PumpStreamHandler(os, es, is) {
            @Override
            public void stop() {
                // when streams need to be stopped, we want to close our input
                try { is.close(); } catch (Exception e) {}
                super.stop();
            }
        };

this.executor
            .command(command)
            .streams(streams);
```

**Comment by reinra** [comment]
I've seen that in case of some Windows processes the streams cannot be closed properly.
You can set ProcessExecutor.closeTimeout() before executing the process to handle this.

**Comment by jjlauer** [comment]
We have a large number of processes being executed quickly, so waiting for a timeout isn't possible.  It would slow down our app by orders of magnitude.  Would you be open to considering an event for when the process itself stops?  It seems strange that in our case, the process stops, but the stream threads are the problem.  They aren't stopped, so the entire stopping process hangs forever.

**Comment by reinra** [comment]
Thanks for pointing out this problem and adding a test for it.
In inputWillPreventCloseWhenProcessStops() I can see that closing the PipedInputStream doesn't unblock anything. Only closing the PipedOutputStream will do it. So user really needs a way of having a custom "closer" in this case. However in you comment you're closing an InputStream instead? Am I missing something?

I found another issue regarding stdin streams that executing two processes in a row and piping stdin to them may actually drop some characters. So I'll take a bigger look into this problem.

**Comment by jjlauer** [comment]
My unit test doesn't really even require a PipedInputStream, I just needed
a stream that mimicked a stream that would be blocking on its read() call.

The key is that its very possible the underlying Process stops, while the
InputStream to that process is valid and blocking on its read().  In that
case, the Thread handling the input will block forever.

**Comment by reinra** [comment]
I pushed the fix for this issue and imported the test you created. Thanks for reporting it.
